php/class-cli.php

```diff
@@ -7,142 +7,238 @@
-	 *
-	 * ## OPTIONS
-	 * [--assets=<assets>]
-	 * : Optional list of assets, comma separated.
-	 *
 	 * ## EXAMPLES
 	 *
 	 *     wp cloudinary sync
 	 *
-	 * @when after_wp_load
+	 * @when    after_wp_load
+	 * @since   2.5.1
 	 *
 	 * @param array $args       Ignored.
-	 * @param array $assoc_args Passed parameters.
+	 * @param array $assoc_args Ignored.
+	 *
+	 * @return void
 	 */
 	public function sync( $args, $assoc_args ) {
 
-		\WP_CLI::log( '' );
-		\WP_CLI::log( '╔═╗┬  ┌─┐┬ ┬┌┬┐┬┌┐┌┌─┐┬─┐┬ ┬  ╔═╗╦  ╦' );
-		\WP_CLI::log( '║  │  │ ││ │ ││││││├─┤├┬┘└┬┘  ║  ║  ║' );
-		\WP_CLI::log( '╚═╝┴─┘└─┘└─┘─┴┘┴┘└┘┴ ┴┴└─ ┴   ╚═╝╩═╝╩' );
+		// Check if analyzed first.
+		if ( empty( get_option( '_cld_cli_analyzed' ) ) ) {
+			$this->analyze();
+		}
+
+		// Initial Query.
+		$query_args = $this->base_query_args;
+		// phpcs:ignore WordPress.DB.SlowDBQuery
+		$query_args['meta_query'] = array(
+			'AND',
+			array(
+				'key'     => '_cld_unsynced',
+				'compare' => 'EXISTS',
+			),
+		);
 
-		$assets = null;
-		if ( ! empty( $assoc_args['assets'] ) ) {
-			$assets = explode( ',', $assoc_args['assets'] );
-			$assets = array_map( 'trim', $assets );
+		// Get assets that need to be synced.
+		$query = new \WP_Query( $query_args );
+		$this->do_process( $query, 'sync' );
+		if ( ! $query->have_posts() ) {
+			\WP_CLI::log( \WP_CLI::colorize( '%gAll assets synced.%n' ) );
 		}
-		$assets = $this->get_all_unsynced( $assets );
 
-		if ( empty( $assets ) ) {
+	}
+
+	/**
+	 * Analyze assets with Cloudinary.
+	 * ## EXAMPLES
+	 *
+	 *     wp cloudinary analyze
+	 *
+	 * @when    after_wp_load
+	 * @since   2.5.1
+	 *
+	 * @return void
+	 */
+	public function analyze() {
+
+		// Initial query.
+		$query_args = $this->base_query_args;
+		$query = new \WP_Query( $query_args );
+
+		// Kill all _cld_ related meta.
+		delete_post_meta_by_key( '_cld_unsynced' );
+		delete_option( '_cld_cli_analyzed' );
+
+		// Do process.
+		$this->do_process( $query, 'analyze' );
+	}
+
+	/**
+	 * Do a process on the query.
+	 *
+	 * @since   2.5.1
+	 *
+	 * @param \WP_Query $query   The initial query object.
+	 * @param string    $process The process to do.
+	 */
+	protected function do_process( &$query, $process ) {
+		// Bail early.
+		if ( ! method_exists( $this, "process_{$process}" ) ) {
+			\WP_CLI::log( \WP_CLI::colorize( "%Invalid Process: {$process}.%n" ) );
+
 			return;
 		}
-		$total = count( $assets );
-		\WP_CLI::log( '' );
-		\WP_CLI::log( \WP_CLI::colorize( '%gStarting Sync:%n' ) );
-		$bar = \WP_CLI\Utils\make_progress_bar( 'Syncing ' . $total . ' assets', $total, 10 );
-		foreach ( $assets as $index => $asset ) {
+		if ( method_exists( $this, $process ) ) {
+			// Setup process.
+			$total   = $query->found_posts;
+			$process = "process_{$process}";
+			do {
+				$posts = $query->get_posts();
+				$this->{$process}( $posts, $total );
+
+				// Free up memory.
+				if ( method_exists( $this, 'stop_the_insanity' ) ) {
+					$this->stop_the_insanity();
```

Comment:
@DavidCramer 

We need to implement this method as it's not part of the WP_CLI.